### PR TITLE
core: directly load infrastructure from RailJSON

### PR DESCRIPTION
**Pull request by multun**
Fix part of https://github.com/OpenRailAssociation/osrd/issues/7056

First commit is a rebased squash of all PRs previously targeting `core/infra_load_refactor` branch.

(some tests are still using legacy builder)

### CODE REVIEW

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+
+            if (detectors.size != 0) {
+                val dirDetector =
+                    if (endpoint == Endpoint.START) detectors[0].decreasing
+                    else detectors[detectors.size - 1].increasing
+                builder.setNextZone(dirDetector, zoneIdx)
+            }
+        }
+
+        // for each detector pair inside the track section
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: This sentence seems incomplete

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+) {
+    val trackDetectors =
+        trackSectionNameToDistanceSortedDetectors.computeIfAbsent(rjsDetector.track) { TreeMap() }
+    val signalPosition = Offset<TrackSection>(rjsDetector.position.meters)
+    val detectorNames = trackDetectors.computeIfAbsent(signalPosition) { mutableListOf() }
+    detectorNames.add(rjsDetector.id)
+}
+
+/** Build detection zones on a partially-filled builder (track-sections, nodes and detectors) */
+private fun buildZones(builder: RawInfraFromRjsBuilder) {
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: This whole function is quite long, and it's not immediately clear to me what every step does (or rather, why)

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt [hunk 3]
```diff
@@ -0,0 +1,756 @@
...
+            uf.union(startEndpoint.index.toInt(), endEndpoint.index.toInt())
+        }
+    }
+
+    // now that union-find is complete, create detection zones
+    val rootToZoneMap = mutableMapOf<Int, ZoneId>()
+    for (trackSectionIdx in builder.getTrackSections()) {
+        val detectors = builder.getTrackSectionDetectors(trackSectionIdx)
+
+        // prepare "node" detection zones
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: What are `"node" detection zones`?

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt [hunk 4]
```diff
@@ -0,0 +1,756 @@
...
+                electrificationRange.trackSectionID,
+                electrificationRange.begin.meters,
+                electrificationRange.end.meters,
+                applyElectrificationForChunkBetween
+            )
+        }
+    }
+
+    // Parse neutral-sections
+    fun parseNeutralRanges(
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: We seem to suddenly switch to use local function half way through `parseRJSInfra()`...

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt [hunk 5]
```diff
@@ -0,0 +1,756 @@
...
+                    }
+
+                    chunkDirNeutralSections.put(
+                        dirChunkLower,
+                        dirChunkUpper,
+                        incomingNeutralSection
+                    )
+
+                    // TODO: remove this priority given to announcement.
+                    // Only added to match previous load, but no functional need exists
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]:
```suggestion
                    // Only added to match the former import process, but no functional need exists
```

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt [hunk 6]
```diff
@@ -0,0 +1,756 @@
...
+                Offset(opPart.position.meters)
+            )
+        }
+    }
+
+    // parse nodes
+    val switchTypeMap =
+        ((rjsInfra.switchTypes ?: listOf()) + RJSSwitchType.BUILTIN_NODE_TYPES_LIST).associateBy {
+            switchType ->
+            switchType.id
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: if the formatter is going to behave this badly, just use `it.id` ?

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+        EdgeDirection.START_TO_STOP -> INCREASING
+        EdgeDirection.STOP_TO_START -> DECREASING
+    }
+}
+
+fun EdgeEndpoint.toEndpoint(): Endpoint {
+    return if (this == EdgeEndpoint.BEGIN) Endpoint.START else Endpoint.END
+}
+
+fun parseRJSInfra(rjsInfra: RJSInfra): RawInfra {
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: Damn, 430 lines long...

We probably should break it up. Not necessarily in this PR, but we should

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt [hunk 8]
```diff
@@ -0,0 +1,822 @@
...
+                                    mapOf(),
+                                    mapOf()
+                                )
+                            )
+                        )
+                    )
+                )
+            )
+        )
+    }
```
Comment (Khoyo) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt]: Huge nit, but this is a 50 lines long statement going ~8 levels of indentation deep

**Review by Khoyo — APPROVED** [review]
Overall LGTM.

Most of my comments are kinda nitpicky, I'll leave to @multun judgement whether to address them now or later.

The process is globally much more readable than the legacy import, but it still feels kinda hard to follow the big picture. I wonder if some restructuring - or more documentation could help.

It's also unclear why some stuff is done in the infra implementation, in the builder and in the parse function - I'm guessing this is still kinda WIP :)

:shipit:

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt (hunk 3 above)
Comment (multun) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: the algorithm is correct but weird, I'll refactor it

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt (hunk 2 above)
Comment (multun) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt]: I've added steps in an attempt to make it clearer in a separate PR : https://github.com/OpenRailAssociation/osrd/pull/7280

**Review by Khoyo — APPROVED** [review]

**Review by leovalais — APPROVED** [review]

File: core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt [hunk 9]
```diff
@@ -0,0 +1,788 @@
...
+                                "[$chunkLower, $chunkUpper]: " +
+                                "${previousNeutralSection.value} != $incomingNeutralSection"
+                        )
+                        break
+                    }
+                }
+
+                chunkDirNeutralSections.put(dirChunkLower, dirChunkUpper, incomingNeutralSection)
+
+                // TODO: remove this priority given to announcement.
```
Comment (bougue-pe) [inline core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt:456]: (it's OK but useless) We have to remove this, too.

Comment (multun) [reply core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt:456]: I suggest we do it in a separate PR, to ensure this one changes as little behavior as possible

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt [hunk 10]
```diff
@@ -0,0 +1,787 @@
...
+    routeEntry: DirDetectorId,
+    exit: DetectorId,
+    nodeConfigs: Map<TrackNodeId, TrackNodeConfigId>,
+    releaseDetectors: StaticIdxSortedSet<Detector>,
+): RouteId {
+    return this.route(routeName) routeBuilder@{
+        // at this point, we have trackRangeIndex pointing to the track range which contains the
+        // entry
+        // detector
+        // we need iterate on track range until we reach the track range which contains the end
```
Comment (Erashin) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt]:
```suggestion
        // we need to iterate on the track range until we reach the track range containing the end
```

**Review by Erashin — APPROVED** [review]
I globally agree with @younesschrifi. This is a very good step (and an essential one, goodbye java infra) towards the right direction. Good job :)

**Review by eckter — APPROVED** [review]
LGTM, I skimmed through the first commit but I did review large parts of it

I tested on STDCM-related stuff, there's a new bug but it's not related to this specific PR (something with conditional speed limits)

I had some weird stuff with gradle / intellij at first when moving to this branch, something about broken cache files. But it fixed itself before I could figure out what was going on.

**Review by clarani — APPROVED** [review]
LGTM, tested :white_check_mark:
